github/resource_github_app_installation_repositories.go

```diff
@@ -0,0 +1,186 @@
+package github
+
+import (
+	"context"
+	"log"
+	"strconv"
+
+	"github.com/google/go-github/v48/github"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+)
+
+func resourceGithubAppInstallationRepositories() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceGithubAppInstallationRepositoriesCreateOrUpdate,
+		Read:   resourceGithubAppInstallationRepositoriesRead,
+		Update: resourceGithubAppInstallationRepositoriesCreateOrUpdate,
+		Delete: resourceGithubAppInstallationRepositoriesDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"installation_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"selected_repositories": {
+				Type: schema.TypeSet,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+				Set:      schema.HashString,
+				Required: true,
+			},
+		},
+	}
+}
+
+func resourceGithubAppInstallationRepositoriesCreateOrUpdate(d *schema.ResourceData, meta interface{}) error {
+	installationIDString := d.Get("installation_id").(string)
+	selectedRepositories := d.Get("selected_repositories")
+
+	client := meta.(*Owner).v3client
+	owner := meta.(*Owner).name
+	ctx := context.WithValue(context.Background(), ctxId, installationIDString)
+
+	selectedRepositoryNames := []string{}
+
+	names := selectedRepositories.(*schema.Set).List()
+	for _, name := range names {
+		selectedRepositoryNames = append(selectedRepositoryNames, name.(string))
+	}
+
+	currentReposNameIDs, instID, err := getAllAccessibleRepos(meta, installationIDString)
+	if err != nil {
+		return err
+	}
+
+	// Add repos that are not in the current state on GitHub
+	for _, repoName := range selectedRepositoryNames {
+		if _, ok := currentReposNameIDs[repoName]; ok {
+			// If it already exists, remove it from the map so we can delete all that are left at the end
+			delete(currentReposNameIDs, repoName)
+		} else {
+			repo, _, err := client.Repositories.Get(ctx, owner, repoName)
+			if err != nil {
+				return err
+			}
+			repoID := repo.GetID()
+			log.Printf("[DEBUG]: Adding %v:%v to app installation %v", repoName, repoID, instID)
+			_, _, err = client.Apps.AddRepository(ctx, instID, repoID)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	// Remove repositories that existed on GitHub but not selectedRepositories
+	// There is a github limitation that means we can't remove the last repository from an installation.
+	// Therefore, we skip the first and delete the rest. The app will then need to be uninstalled via the GUI
+	// as there is no current API endpoint for [un]installation. Ensure there is at least one repository remaining.
```

Comment:
As per the comment here, I skip the first... which is base on the order of repositories that are returned by the the [list accessible repos API call](https://docs.github.com/en/rest/apps/installations?apiVersion=2022-11-28#list-repositories-accessible-to-the-user-access-token).
I guess I could sort it alphabetically so that it is always the first alphabetical by name that remains?

Reply:
This seems like a reasonable call. It's definitely an instance where GitHub's behavior doesn't fit a tidy CRUD model.